vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc

```diff
@@ -813,6 +825,16 @@ void AdsImpl::OnAdNotificationClicked(const AdNotificationInfo& ad) {
 
   epsilon_greedy_bandit_processor_->Process(
       {ad.segment, mojom::AdNotificationEventType::kClicked});
+
+  auto ad_notification_clicked_covariate_log_entry =
+      std::make_unique<AdNotificationClickedCovariateLogEntry>();
+  ad_notification_clicked_covariate_log_entry->SetClicked(true);
+  CovariateLogs::Get()->SetCovariateLogEntry(
+      std::move(ad_notification_clicked_covariate_log_entry));
+
+  mojom::TrainingInstancePtr instance =
+      CovariateLogs::Get()->GetTrainingInstance();
+  AdsClientHelper::Get()->LogTrainingCovariates(std::move(instance));
```

Comment:
We could move this to a `training_covariates_util.h/.cc` with a `LogTrainingCovariates()` free method as this way we just need to call to `LogTrainingCovariates()` to remove duplicate code

Reply:
It would also appear that the above is always called after `CovariateLogs::Get()->SetCovariateLogEntry`, could we add the setter inside the same util method, and then call `LogTrainingCovariates(AnyCovariateLogEntry)` which would essentially replace:
```
  auto ad_notification_clicked_covariate_log_entry =
      std::make_unique<AdNotificationClickedCovariateLogEntry>();
  ad_notification_clicked_covariate_log_entry->SetClicked(true);
  CovariateLogs::Get()->SetCovariateLogEntry(
      std::move(ad_notification_clicked_covariate_log_entry));

  mojom::TrainingInstancePtr instance =
      CovariateLogs::Get()->GetTrainingInstance();
  AdsClientHelper::Get()->LogTrainingCovariates(std::move(instance));
```
with
```
  auto ad_notification_clicked_covariate_log_entry =
      std::make_unique<AdNotificationClickedCovariateLogEntry>();
  ad_notification_clicked_covariate_log_entry->SetClicked(true);
  LogTrainingCovariate(ad_notification_clicked_covariate_log_entry);
```

Reply:
ditto for other areas of the code

Reply:
@tmancey as discussed with @aseren w/o globals this makes more sense inside the covariate logs class. Wdyt?

Reply:
Sounds great to me